File: src/main/java/simple/nodes/exps/VarNode.java
package simple.nodes.exps;

import com.oracle.truffle.api.frame.FrameDescriptor;
import simple.Environment;
import simple.nodes.Node;
import truffle.nodes.TFNode;
import truffle.parser.LexicalScope;

import java.util.Objects;

public record VarNode(String name) implements Node {
    @Override
    public Object eval(Environment env) {
        return env.getValue(name);
    }

    @Override
    public TFNode convert(FrameDescriptor.Builder descriptorBuilder, LexicalScope scope) {
        return null;
    }

    @Override
    public boolean equals(Object another) {
        if (another instanceof VarNode) {
            return Objects.equals(this.name, ((VarNode) another).name);
        }

        return false;
    }
}

File: src/main/kotlin/parser/expressions/VarNode.kt
package parser.expressions

import com.oracle.truffle.api.frame.FrameSlot
import com.oracle.truffle.api.frame.VirtualFrame

class VarNode(private val slot: FrameSlot) : PrimeExp() {
    /*override fun execute(virtualFrame: VirtualFrame): Int {
        return virtualFrame.getInt(index)
    }*/

    override fun execute(virtualFrame: VirtualFrame): Int {
        return virtualFrame.getInt(slot)
    }
}
